tempesta_fw/apm.c

```diff
@@ -825,31 +833,42 @@ tfw_apm_prcntl_fn(unsigned long fndata)
  *
  * Return 0 if the percentile values didn't need recalculation.
  * Return 1 if potentially new percentile values were calculated.
+ *
+ * The two functions below differ only by the type of lock used.
+ * tfw_apm_stats() should be used for calls in kernel context.
+ * tfw_apm_stats_bh() should be used for calls in user context.
  */
+#define __tfw_apm_stats_body(apmdata, pstats, fn_lock, fn_unlock)	\
+	int rdidx, seq = pstats->seq;					\
+	TfwApmData *data = apmdata;					\
+	TfwApmSEnt *asent;						\
+									\
+	BUG_ON(!apmdata);						\
+									\
+	smp_mb__before_atomic();					\
+	rdidx = (unsigned int)atomic_read(&data->stats.rdidx) % 2;	\
+	asent = &data->stats.asent[rdidx];				\
+									\
+	fn_lock(&asent->rwlock);					\
+	memcpy(pstats->val, asent->pstats.val,				\
+	       pstats->psz * sizeof(pstats->val[0]));			\
+	fn_unlock(&asent->rwlock);					\
+	pstats->seq = rdidx;						\
+									\
+	return (seq != rdidx);
```

Comment:
Is there a way to know whether the actual values were changed? E.g there can be 1000 of very busy servers, but after some time they go to some "stable" state, when their response times remain *almost* the same. "Almost" is crucial, because if a server response time changes for exampe for 1/100 or even 1/1000, then there is no sense to trigger upper LB logic about the irrelevant changes.

Meantime, real life farms of 1000 servers are usually always have issues with several servers, so there the real (probably) case is that 99% of servers are in the "stable" state, while the rest of the servers face issues and their weights must be adjusted. In sense of #76 it would be good to be able to catch the exact set of servers with changed pstats.